[16.0][MIG] delivery_carrier_package_measure_required: Migration to 16.0

RodrigoBM opened the pull request:
issue #534 

Migration from 14.0

TDu committed:
Add delivery_carrier_package_measure_required

This module adds the possibility to flag some measurements on package to be
required for delivery.

TDu committed:
Fix dc_package_measure_required for not done transfer

The required dimensions on a package need not to be enforce before the
picking is being done (package with content).
Because that blocks any preparation on the package before it is
delivered.

But we still want to be able to check if the package is valid at will.
This can be done with the `delivery_pkg_measure__ignore_package_content`
context key.

OCA-git-bot committed:
delivery_carrier_package_measure_required 14.0.1.0.1

TDu committed:
dc_package_measure_required: fix warning constrains

OCA-git-bot committed:
delivery_carrier_package_measure_required 14.0.1.0.2

FranciscoFactorLibre committed:
Added translation using Weblate (Spanish)

RodrigoBM committed:
[IMP] delivery_carrier_package_measure_required: pre-commit stuff

RodrigoBM committed:
[MIG] delivery_carrier_package_measure_required: Migration to 16.0"

xAdrianCif commented on delivery_carrier_package_measure_required/tests/test_delivery_carrier_package_measure_require.py:
```diff
@@ -0,0 +1,107 @@
+                "carrier_id": cls.normal_carrier.id,
+            }
+        )
+        cls.env["stock.move.line"].create(
+            {
+                "product_id": cls.product_aw.id,
+                "product_uom_id": cls.uom_kg.id,
+                "picking_id": cls.pick.id,
+                "qty_done": 5,
+                "location_id": cls.stock_location.id,
+                "location_dest_id": cls.customer_location.id,
+            }
+        )
+        cls.pick.action_confirm()
```
I know that this is a migration and comes from there, but to respect the stock flow in Odoo, it's better to create a `stock.move` (instead of a `stock.move.line`) directly associated with `stock.picking`, and then force availability using the 'action_assign' method. If there is availability, the `stock.move.line` is created, and we can modify its `qty_done`.

RodrigoBM replied:
done

xAdrianCif reviewed (approved):
Tech review, thanks for the job!

RodrigoBM commented:
is it possible to merge this, @OCA/logistics-maintainers, thank you

RodrigoBM commented:
ping @TDu  @rousseldenis @simahawk @jbaudoux @lmignon

jorgemartinez-factorlibre reviewed (approved):
Tech review, good job @RodrigoBM

hugosantosred commented:
@OCA/logistics-maintainers Could you review for merge or apply the no stale label to this PR? Thanks!

simahawk commented:
@RodrigoBM can you rebase pls?

simahawk reviewed (commented):
@TDu ping for review ;)

RodrigoBM commented:
> @RodrigoBM can you rebase pls?

Done @simahawk

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -19,3 +19,23 @@ def _check_required_package_measurement(self):
         self._check_required_package_measurement()
         return super().button_validate()
+
+    def _put_in_pack(self, move_line_ids, create_package_level=True):
+        res = super()._put_in_pack(
+            move_line_ids, create_package_level=create_package_level
+        )
+        package_length = self._context.get("choose_delivery_package_length", 0)
+        package_width = self._context.get("choose_delivery_package_width", 0)
+        package_height = self._context.get("choose_delivery_package_height", 0)
+        package_weight = self._context.get("choose_delivery_package_pack_weight", 0)
+
+        res.write(
+            {
+                "pack_length": package_length,
+                "width": package_width,
+                "height": package_height,
+                "pack_weight": package_weight,
+            }
+        )
+
+        return res
```
I guess not part of the migration but an improvement. If yes, can it go in a separate commit?

RodrigoBM replied:
Done.

RodrigoBM replied:
this is now resolved, @jbaudoux?

jbaudoux commented on delivery_carrier_package_measure_required/models/stock_quant_package.py:
```diff
@@ -8,20 +8,23 @@
         )
         for package in self:
-            if not ignore_pack_content and not package.quant_ids:
+            if ignore_pack_content or (not force_validation and not package.quant_ids):
```
I think we better stick to previously existing context variable (and not invert behavior). I don't find it clearer with the new name.

RodrigoBM replied:
@jbaudoux , I think this is better now.

Before the changes we don't check the dimensions in the case if ignore_pack_content is False and the package has no quants.
We also have a variable that is passed by context which is delivery_pkg_measure__ignore_package_content that if you pass it with the value set to True you would expect it not to check the dimensions but this is not the case.

Option1:
ignore_pack_content = True, package.quant_ids = True
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 2:
ignore_pack_content = True, package.quant_ids = False
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 3:
ignore_pack_content = False, package.quant_ids = True
(not ignore_pack_content and not package.quant_ids) if == False --> Check dimensions
Opción 4:
ignore_pack_content = False, package.quant_ids = False
(not ignore_pack_content and not package.quant_ids) if == True --> No Check dimensions

But what if we want to be able to skip the dimension check in other cases?

Option1:
ignore_pack_content = False, force_validation = True, package.quant_ids = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 2:
ignore_pack_content = False, force_validation = True, package.quant_ids = False
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 3:
ignore_pack_content = False, force_validation = False, package.quant_ids = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = False --> Check dimensions
Opción 4:
ignore_pack_content = False, force_validation = False, package.quant_ids = False
ignore_pack_content or (not force_validation and not package.quant_ids) if = True --> No Check dimensions
Opción 5:
ignore_pack_content = True
ignore_pack_content or (not force_validation and not package.quant_ids) if = True --> No Check dimensions

What do you think about this? this can go in another commit if you think it is more correct.

RodrigoBM replied:
however, the delivery_pkg_measure__ignore_package_content context I have not seen used in other OCA modules.

jbaudoux replied:
I can't find any usage of this context variable. @TDu What was the purpose of it?
Do we need any of those context variables? If this is useless, I'm more in favor of simplifying. We can already configure which dimension is required.

RodrigoBM replied:
I understand @jbaudoux , but now I have client developments that need these contexts. We can change this "if ignore_pack_content or (not force_validation and not package.quant_ids):" to "if self.ignore_check_dimensions():" and have everyone inherit the method to whatever we need.

RodrigoBM replied:
ping @TDu

TDu replied:
As commented in the code that content key `delivery_pkg_measure__ignore_package_content`  was meant to be used to force asking for dimension even for empty packages.

I think the idea behind it was, you receive a new product in your warehouse, all packaging (A, B) for that product are not configured properly. The product arrives in packaging A. But you want the user to configure all packaging (A and B) for that product on first reception.

Now indeed it does not look like it is used :thinking:

RodrigoBM replied:
Is it a problem to accept this new proposal? Do you need it to be in a separate commit?

jbaudoux replied:
Thanks for the clarification @TDu 
@RodrigoBM Your current proposal is fine for me. It is clear and covers all the cases.
Just put that in a separate commit :)

RodrigoBM replied:
Done @jbaudoux, sorry for the delay

jbaudoux commented:
/ocabot migration delivery_carrier_package_measure_required

RodrigoBM committed:
[IMP] delivery_carrier_package_measure_required: put in pack

add measures to package when you add producto to pack.

TDu commented on delivery_carrier_package_measure_required/models/stock_picking.py:
```diff
@@ -0,0 +1,41 @@
+    def _check_required_package_measurement(self):
+        """Check the required measurement on destination pacakges.
+
+        If package is None all result packages on the stock pickings are checked
+        otherwise only the one
```
```suggestion
```
docstring out of sync with the code ?

RodrigoBM replied:
Now that I think about it, I think so

RodrigoBM replied:
Done

RodrigoBM replied:
this is now resolved, @TDu?

RodrigoBM commented:
ping @TDu @jbaudoux

RodrigoBM committed:
[IMP] delivery_carrier_package_measure_required: dimension control exceptions

greater granularity is implemented to control possible dimension control exceptions.

jbaudoux commented:
@RodrigoBM As a general rule, can you keep commit subject (first line) short and put description in the body (separated by a blank line from the subject).

Like:
```
delivery_carrier_package_measure_required: put in pack

Add measures to package when you add product to pack
```

RodrigoBM commented:
> @RodrigoBM As a general rule, can you keep commit subject (first line) short and put description in the body (separated by a blank line from the subject).
> 
> Like:
> 
> ```
> delivery_carrier_package_measure_required: put in pack
> 
> Add measures to package when you add product to pack
> ```

Done. I will take this tip in consideration in the next contributions.

jbaudoux reviewed (approved):
Thanks

jbaudoux commented:
/ocabot merge nobump
